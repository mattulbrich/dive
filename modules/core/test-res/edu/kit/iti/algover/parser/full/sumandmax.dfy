// RUN: %dafny /compile:0 "%s" > "%t"
// RUN: %diff "%s.expect" "%t"

// VSComp 2010, problem 1, compute the sum and max of the elements of an array and prove
// that 'sum <= N * max'.
// Rustan Leino, 18 August 2010.
//
// The problem statement gave the pseudo-code for the method, but did not ask to prove
// that 'sum' or 'max' return as the sum and max, respectively, of the array.  The
// given assumption that the array's elements are non-negative is not needed to establish
// the requested postcondition.
//// class C { var f:int;} class D { var g:object; }
// ALGOVER SIMP: removed in contract clauses;
method M(N: int, a: array<int>) returns (sum: int, max: int)
  requires 0 <= N && a != null && a.Length == N && (forall k : int :: 0 <= k && k < N ==> 0 <= a[k])
  ensures sum <= N * max
{
  sum := 0;
  max := 0;
// ALGOVER SIMP: var i := 0;
  var i : int := 0;
  while (i < N)
    invariant i <= N && sum <= i * max
  {
    if max < a[i]
    {
      max := a[i];
    }
    sum := sum + a[i];
    i := i + 1;
  }
}

method Main()
{
  // ALGOVER SIMPL var a := new int[10];
  var a : array<int>; //:= new int[10];
  a[0] := 9;
  a[1] := 5;
  a[2] := 0;
  a[3] := 2;
  a[4] := 7;
  a[5] := 3;
  a[6] := 2;
  a[7] := 1;
  a[8] := 10;
  a[9] := 6;
  // ALGOVER SIMP: var s, m := M(10, a);
  var s : int;
  var m : int;
  // s, m := M(10,a);
  // ALGOVER SIMP: print "N = ", a.Length, "  sum = ", s, "  max = ", m, "\n";
}

