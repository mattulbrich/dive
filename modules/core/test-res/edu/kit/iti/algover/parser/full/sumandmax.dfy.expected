
(COMPILATION_UNIT 
 (method M 
  (ARGS 
   (VAR N int) 
   (VAR a 
    (array int))) 
  (returns 
   (VAR sum int) 
   (VAR max int)) 
  (requires 
   (&& 
    (&& 
     (&& 
       (<= 0 N) 
       (!= a null))
      (== (Length a) N))
      (forall k int 
       (==> 
        (&& 
         (<= 0 k) 
         (< k N)) 
        (<= 0 
         (ARRAY_ACCESS a k)))))) 
  (ensures 
   (<= sum 
    (* N max))) 
  (BLOCK 
   (:= sum 0) 
   (:= max 0) 
   (var i int 0) 
   (while 
    (< i N) 
    (invariant 
     (&& 
      (<= i N) 
      (<= sum 
       (* i max)))) 
    (BLOCK 
     (if 
      (< max 
       (ARRAY_ACCESS a i)) 
      (BLOCK 
       (:= max 
        (ARRAY_ACCESS a i)))) 
     (:= sum 
      (+ sum 
       (ARRAY_ACCESS a i))) 
     (:= i 
      (+ i 1)))))) 
 (method Main ARGS 
  (BLOCK 
   (var a 
    (array int)) 
   (:= 
    (ARRAY_ACCESS a 0) 9) 
   (:= 
    (ARRAY_ACCESS a 1) 5) 
   (:= 
    (ARRAY_ACCESS a 2) 0) 
   (:= 
    (ARRAY_ACCESS a 3) 2) 
   (:= 
    (ARRAY_ACCESS a 4) 7) 
   (:= 
    (ARRAY_ACCESS a 5) 3) 
   (:= 
    (ARRAY_ACCESS a 6) 2) 
   (:= 
    (ARRAY_ACCESS a 7) 1) 
   (:= 
    (ARRAY_ACCESS a 8) 10) 
   (:= 
    (ARRAY_ACCESS a 9) 6) 
   (var s int) 
   (var m int))))

